Add postgresql upgrade script

carbonin opened the pull request:
This script upgrades PostgreSQL between major versions.

It takes care of changing the mount point of our appliance's data disk from one version to the other, while also upgrading the data without using extra space and without a dump/restore by using hard links (`-k` option to `pg_upgrade`).

This script will need to change slightly as we change versions of PostgreSQL as right now, it also moves the mount point of the data volume from the data directory to the `pgsql` directory.

This was accomplished on new appliances in the following three PRs:
https://github.com/ManageIQ/manageiq/pull/11597
https://github.com/ManageIQ/manageiq-appliance/pull/91
https://github.com/ManageIQ/manageiq-appliance-build/pull/165

This script should be run as a part of the in-place upgrade procedure after the new manageiq code and new PostgreSQL RPM have been installed.

After the upgrade has run, the `rh-postgresql94-postgresql` rpms can be removed.

@Fryguy @gtanzillo @simaishi please review

carbonin committed:
Add a script to upgrade PostgreSQL

This script will need to change slightly as we change versions of
PostgreSQL

It takes care of changing the mountpoint of our appliance's data
disk from one version to the other, while also upgrading the data
without using extra space and without a dump/restore.

carbonin committed:
Correct fstab sed command

Previously it was changing the mountpoint version, but not removing
the /data from the path.

This change also uses `mount -a` rather than an explicit mount command
to be sure fstab is being used.

carbonin committed:
Disable PostgreSQL 9.4 after upgrade

carbonin commented:
/cc @fbladilo this may be interesting for the PostgreSQL container deployment as well. I'm not sure how that will work though as this requires both versions of PostgreSQL to be installed at the time of the upgrade.

Fryguy commented on LINK/usr/bin/miq_postgres_upgrade.sh:
```diff
@@ -0,0 +1,147 @@
+fi
+
+# make sure neither postgres service is running
+systemctl status ${OLD_PG_SERVICE} &> /dev/null
```
We should just call stop and wait for it to stop

Fryguy commented on LINK/usr/bin/miq_postgres_upgrade.sh:
```diff
@@ -0,0 +1,147 @@
+ -shared_preload_libraries = 'pglogical'   # MIQ Value (change requires restart)
+ +shared_preload_libraries = 'pglogical,repmgr_funcs'   # MIQ Value (change requires restart)
+
+ +wal_log_hints = on
```
It would be nice if we had a patch, applied it, and if it failed to apply, then show this message.  That way those customers that have made no changes would not have any manual intervention

Fryguy commented:
As discussed,
- You might want to add `set -e`
- We may want to implement some sort of migration process, so that as new versions come out, and a user jumps from, say, 9.4 to 9.6, we would know to run the 9.4->9.5, then 9.5->9.6 migration files.
- I can't remember the third thing 😝

carbonin committed:
Add `set -e` so that the script exits on failure

carbonin committed:
Stop the postgres services rather than making the user do it

carbonin committed:
Ignore errors from the mv command for changing the mount point

We expect an error here because you can't move the data directory
into itself, but that's okay for our usecase.

carbonin commented:
@Fryguy I made some of the changes. I left out the patch and version changes as they will probably require more work to do properly.

I would like to get this in sooner rather than later so we can start testing end-to-end upgrades.
I'll do the other enhancements in followups if that's okay with you.

carbonin commented:
@Fryguy , @gtanzillo any more feedback on this?

fbladilo commented:
@carbonin With a few environment modifications, the script could potentially be used on containerized builds, most of them are related to the fact that containers do not count with direct access to underlying storage (LVM), only persistent volumes are presented to them.

There is also a pending decision if the upstream centos postgresql image will officially support upgrades, even if this does not happen, we could prepare an image with dual PG versions only for the upgrade/transition.

carbonin commented:
@fbladilo Okay, then we may need a different approach for containers for now.

I have some changes that @Fryguy mentioned that would make this more generally usable so one part of that might be moving the mount point. Then we can make that conditional on whether we are a container or not (hopefully). Or you could just reference the functions I create for doing the upgrade and have a separate base script.

Either way it looks like we will probably be separate for this upgrade.

fbladilo commented:
@carbonin 👍 I was thinking something along those lines as well, we have enough environment to detect container based builds and inject a conditional with the upgrade logic we need for container environments.

chessbyte commented:
Euwe Backport details:

``` sh
$ git log -1
commit a145cf6ec2e51744014fe6d41ebc24ea26a76f49
Author: Jason Frey <fryguy9@gmail.com>
Date:   Fri Oct 14 11:56:39 2016 -0400

    Merge pull request #92 from carbonin/add_postgresql_upgrade_script

    Add postgresql upgrade script
    (cherry picked from commit 622e1f5bb0e0bab9a0283799f8c1bfaacb0ee689)
```
